examples/move/big_vector/sources/big_vector.move

```diff
@@ -0,0 +1,1274 @@
+            max_slice_size: _,
+            max_fan_out: _,
+
+            root_id: _,
+            last_id: _,
+        } = self;
+
+        assert!(length == 0, ENotEmpty);
+        id.delete();
+    }
+
+    /// Destroy `self`, even if it contains elements, as long as they
+    /// are droppable.
+    public fun drop<E: store + drop>(self: BigVector<E>) {
+        let BigVector {
+            mut id,
+
+            depth,
+            length: _,
+            max_slice_size: _,
+            max_fan_out: _,
+
+            root_id,
+            last_id: _,
+        } = self;
+
+        drop_slice<E>(&mut id, depth, root_id);
+        id.delete();
+    }
+
+    // === BigVector Accessors ===
+
+    /// Whether `self` contains no elements or not.
+    public fun is_empty<E: store>(self: &BigVector<E>): bool {
+        self.length == 0
+    }
+
+    /// The number of elements contained in `self`.
+    public fun length<E: store>(self: &BigVector<E>): u64 {
+        self.length
+    }
+
+    /// The number of nodes between the root and the leaves in `self`.
+    /// This is within a constant factor of log base `max_fan_out` of
+    /// the length.
+    public fun depth<E: store>(self: &BigVector<E>): u8 {
+        self.depth
+    }
+
+    #[syntax(index)]
+    /// Access the element at index `ix` in `self`.
+    public fun borrow<E: store>(self: &BigVector<E>, ix: u128): &E {
+        let (ref, offset) = self.slice_around(ix);
+        let slice = self.borrow_slice(ref);
+        &slice[offset]
+    }
+
+    #[syntax(index)]
+    /// Access the element at index `ix` in `self`, mutably.
+    public fun borrow_mut<E: store>(self: &mut BigVector<E>, ix: u128): &mut E {
+        let (ref, offset) = self.slice_around(ix);
+        let slice = self.borrow_slice_mut(ref);
+        &mut slice[offset]
+    }
+
+    // === BigVector Mutators ===
+
+    /// Add `val` to `self` at index `key`. Aborts if `key` is already
+    /// present in `self`.
+    public fun insert<E: store>(self: &mut BigVector<E>, key: u128, val: E) {
+        self.length = self.length + 1;
+
+        if (self.root_id == NO_SLICE) {
+            self.root_id = self.alloc(singleton(key, val));
+            return
+        };
+
+        let (root_id, depth) = (self.root_id, self.depth);
+        let (key, other) = self.slice_insert(root_id, depth, key, val);
+
+        if (other != NO_SLICE) {
+            self.root_id = self.alloc(branch(key, root_id, other));
+            self.depth = self.depth + 1;
+        }
+    }
+
+    /// Adds key value pairs from `keys` and `vals` to `self`.
+    /// Requires that `keys` and `vals` have the same length, and that
+    /// `keys` is in sorted order.
+    ///
+    /// Aborts if any of the keys are already present in `self`, or
+    /// the requirements on `keys` and `vals` are not met.
+    public fun insert_batch<E: store>(
+        _self: &mut BigVector<E>,
+        _keys: vector<u128>,
+        _vals: vector<E>,
+    ) {
+        abort 0
```

Comment:
Left unimplemented for now, because it's a little fiddly/complicated, and an implementation without this should still be an improvement on what we have today, but @gdanezis pointed out that having the ability to insert and remove in batches would be a helpful primitive to expose to efficiently handle workloads due to market makers who are pushing in orders in bulk.

Once we get an idea of whether or not this implementation is useful, we can add these extra bells and whistles.

Reply:
Note: to be useful it would have to implement batch_op (where you can do in one go both insert, update, or delete), return an idea of mutations, and not abort (so you can mix many different parties operations). So yes, lets wait and see where it might be useful.

Reply:
Got it, so I was imagining this being used by a market maker who was deleting their old orders and putting in a number of new orders, but it sounds like you have something else in mind, perhaps where you gather updates to the table across multiple transactions into a journal and then flush them in a batch, is that right?

(I have gone slightly off my idea of using a $B^\epsilon$ tree for this, because it would really complicate iteration, otherwise it would work well, but maybe we can get the best of both worlds by literally storing a journal at a single level?)